### PR TITLE
many: implement fetching sections and package names periodically.

### PR DESCRIPTION
**Pull request by chipaca**
These are then used for quicker tab completion.

### CODE REVIEW

File: dirs/dirs.go [hunk 1]
```diff
@@ -187,6 +191,10 @@ func SetRootDir(rootdir string) {
 
 	SnapStateFile = filepath.Join(rootdir, snappyDir, "state.json")
 
+	SnapCacheDir = filepath.Join(rootdir, snappyDir, "cache")
```
Comment (zyga) [inline dirs/dirs.go]: Why not `/var/cache/snapd`?

File: overlord/snapstate/snapmgr.go [hunk 2]
```diff
@@ -511,6 +530,32 @@ func (m *SnapManager) ensureRefreshes() error {
 	return err
 }
 
+// ensureCatalogRefresh ensures that we refresh the cataloge
+// data periodically
+func (m *SnapManager) ensureCatalogRefresh() error {
+	// sneakily don't do anything if in testing
```
Comment (zyga) [inline overlord/snapstate/snapmgr.go]: We could move the mkdir here so that it stays operational even if someone just goes and does `rm -rf` on the thing after snapd started.

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -278,6 +282,9 @@ func Manager(st *state.State) (*SnapManager, error) {
 	if err := os.MkdirAll(dirs.SnapCookieDir, 0700); err != nil {
 		return nil, fmt.Errorf("cannot create directory %q: %v", dirs.SnapCookieDir, err)
 	}
+	if err := os.MkdirAll(dirs.SnapCacheDir, 0755); err != nil {
```
Comment (zyga) [inline overlord/snapstate/snapmgr.go]: Please consider moving this below.

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -1670,3 +1672,26 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
...
+	if err != nil {
+		return err
+	}
+
+	sort.Strings(sections)
+	if err := osutil.AtomicWriteFile(dirs.SnapSectionsFile, []byte(strings.Join(sections, "\n")), 0644, 0); err != nil {
+		return err
+	}
+
+	namesFd, err := osutil.NewAtomicFile(dirs.SnapNamesFile, 0644, 0, -1, -1)
```
Comment (zyga) [inline overlord/snapstate/snapstate.go]: Thank you, much less gray hair today :-)

File: dirs/dirs.go (hunk 1 above)
Comment (chipaca) [reply dirs/dirs.go]: I'd like it to be `/var/cache/snapd` as well, but that would mean adding that to `writable-paths`, and I don't know the impact on packaging elsewhere (is `/var/cache/{package}` universal?); keeping it under `snappyDir` makes it a no-brainer.

**Comment by chipaca** [comment]
Marking this blocked so it isn't merged in error (agreed with server-side about sorting the names endpoint, so we'll wait for that)

File: overlord/snapstate/snapmgr.go (hunk 3 above)
Comment (niemeyer) [reply overlord/snapstate/snapmgr.go]: Below what?

File: store/store.go [hunk 5]
```diff
@@ -1104,6 +1147,42 @@ func (s *Store) Sections(user *auth.UserState) ([]string, error) {
 	return sectionNames, nil
 }
 
+// WriteCommandsCatalogs queries the "commands" endpoint and writes
+// the command names into the given io.Writer [and at some point the
+// CNF catalog to another io.Writer; adding now to minimise interface
+// churn].
+func (s *Store) WriteCommandsCatalogs(names io.Writer, _ io.Writer) error {
```
Comment (niemeyer) [inline store/store.go]: The function name and docs would benefit from some cleaning up after the repurposing of this interface. _WriteCatalogs_ alone sounds pretty reasonable as a name, and we can also drop the second argument while it's missing on _both ends_ of the function call.

File: store/store.go [hunk 6]
```diff
@@ -1104,6 +1147,42 @@ func (s *Store) Sections(user *auth.UserState) ([]string, error) {
...
+		Method: "GET",
+		URL:    &u,
+		Accept: halJsonContentType,
+	}
+
+	resp, err := httputil.RetryRequest(u.String(), func() (*http.Response, error) {
+		return s.doRequest(context.TODO(), s.client, reqOptions, nil)
+	}, func(resp *http.Response) error {
+		return decodeCatalog(resp, names)
+	}, defaultRetryStrategy)
```
Comment (niemeyer) [inline store/store.go]: Quite hard to follow what's going on here. This will probably improve much if we define the arguments ahead of the function call and then have a single line with the call.

File: cmd/snap/complete.go [hunk 7]
```diff
@@ -28,9 +31,39 @@ func (s installedSnapName) Complete(match string) []flags.Completion {
 	return ret
 }
 
+func completeFromSortedFile(fn, match string) ([]flags.Completion, error) {
```
Comment (niemeyer) [inline cmd/snap/complete.go]: s/fn/filename/ wouldn't hurt. This is used a single time, and improves the readability of the function definition.

File: cmd/snap/complete.go [hunk 8]
```diff
@@ -28,9 +31,39 @@ func (s installedSnapName) Complete(match string) []flags.Completion {
...
+	fd, err := os.Open(fn)
```
Comment (niemeyer) [inline cmd/snap/complete.go]: s/fd/file/ or similar

Let's please not get used to calling file objects or readers/writers as descriptors. A file descriptor is a number, obtained in this case via file.Fd()

File: cmd/snap/complete.go [hunk 9]
```diff
@@ -28,9 +31,39 @@ func (s installedSnapName) Complete(match string) []flags.Completion {
...
+		return nil, err
+	}
+	defer fd.Close()
+
+	var ret []flags.Completion
+
+	// TODO: look into implementing binary search
+	//       e.g. https://github.com/pts/pts-line-bisect/
+	scanner := bufio.NewScanner(fd)
+	for scanner.Scan() {
```
Comment (niemeyer) [inline cmd/snap/complete.go]: Probably good enough for ~2k package names, but we should look into a more efficient format for commands or once we get too many items.

File: cmd/snap/complete.go [hunk 10]
```diff
@@ -28,9 +31,39 @@ func (s installedSnapName) Complete(match string) []flags.Completion {
...
+		line := scanner.Text()
+		if line >= match {
+			if strings.HasPrefix(line, match) {
+				ret = append(ret, flags.Completion{Item: line})
+			} else {
+				break
+			}
+		}
```
Comment (niemeyer) [inline cmd/snap/complete.go]: Slightly easier to read:
```
if line < match {
        continue
}
if strings.HasPrefix(line, match) {
        ret = append(ret, ...)
        continue
}
break

File: daemon/api_test.go [hunk 11]
```diff
@@ -288,6 +286,8 @@ func (s *apiBaseSuite) daemon(c *check.C) *Daemon {
 		Serial: "serialserial",
 	})
 
+	snapstate.CanAutoRefresh = nil
```
Comment (niemeyer) [inline daemon/api_test.go]: Why was this moved? Seems much safer to have it very early on while the suite is being setup rather than on a function after much logic has been run, and even the daemon was created.

File: daemon/daemon_test.go [hunk 12]
```diff
@@ -63,10 +62,6 @@ func (s *daemonSuite) checkAuthorizationForPid(pid uint32, actionId string, deta
 	return s.authorized, s.err
 }
 
-func (s *daemonSuite) SetUpSuite(c *check.C) {
-	snapstate.CanAutoRefresh = nil
```
Comment (niemeyer) [inline daemon/daemon_test.go]: Same.

File: tests/main/completion/task.yaml [hunk 13]
```diff
@@ -15,6 +23,7 @@ prepare: |
     expect -d -f key.exp0
 
 restore: |
+    chattr -i /var/cache/snapd/names
```
Comment (niemeyer) [inline tests/main/completion/task.yaml]: Shouldn't it be removed since prepare created it?

File: dirs/dirs.go [hunk 14]
```diff
@@ -187,6 +191,10 @@ func SetRootDir(rootdir string) {
 
 	SnapStateFile = filepath.Join(rootdir, snappyDir, "state.json")
 
+	SnapCacheDir = filepath.Join(rootdir, "/var/cache/snapd")
```
Comment (niemeyer) [inline dirs/dirs.go]: We need to fix packaging so there's proper ownership on this directory on all distributions, and the packaging purging logic for all distros also need to account for this new directory.

File: overlord/snapstate/snapmgr.go (hunk 3 above)
Comment (chipaca) [reply overlord/snapstate/snapmgr.go]: below, where his other comment is (or was). I know what he means (and i forgot to do it in my last round)

File: dirs/dirs.go (hunk 14 above)
Comment (zyga) [reply dirs/dirs.go]: :heart: - indeed

File: overlord/snapstate/snapmgr.go [hunk 15]
```diff
@@ -526,6 +545,32 @@ func (m *SnapManager) ensureRefreshes() error {
...
 }
 
+// ensureCatalogRefresh ensures that we refresh the catalog
+// data periodically
+func (m *SnapManager) ensureCatalogRefresh() error {
+	// sneakily don't do anything if in testing
+	if CanAutoRefresh == nil {
+		return nil
+	}
+	m.state.Lock()
```
[the remark this answers is not in the document]
Comment (chipaca) [reply overlord/snapstate/snapmgr.go]: I need to hold the lock to get the cached store object, and to set the next timeout; I need it unlocked to call refreshCatalogs. So it'd have to be
```golang
m.state.Lock()
defer m.state.Unlock()
// do stuff
m.state.Unlock()
defer m.state.Lock()
// do other stuff
```
I thought it best to avoid the resulting spurious lock/unlock defers

File: overlord/snapstate/snapmgr.go [hunk 16]
```diff
@@ -526,6 +545,32 @@ func (m *SnapManager) ensureRefreshes() error {
...
+	aStore := storestate.Store(m.state)
```
[the remark this answers is not in the document]
Comment (chipaca) [reply overlord/snapstate/snapmgr.go]: grep agrees with you

File: store/store.go [hunk 17]
```diff
@@ -459,6 +460,7 @@ func New(cfg *Config, authContext auth.AuthContext) *Store {
 		store.buyURI = endpointURL(cfg.StoreBaseURL, "api/v1/snaps/purchases/buy", nil)
 		store.customersMeURI = endpointURL(cfg.StoreBaseURL, "api/v1/snaps/purchases/customers/me", nil)
 		store.sectionsURI = endpointURL(cfg.StoreBaseURL, "api/v1/snaps/sections", nil)
+		store.commandsURI = endpointURL(cfg.StoreBaseURL, "api/v1/snaps/names", nil)
```
[the remark this answers is not in the document]
Comment (chipaca) [reply store/store.go]: it's called names, but it's (going to be) a list of commands and aliases per package. I thought it best to call it for what it is, rather than for what its url says it is.

File: store/store.go (hunk 5 above)
Comment (chipaca) [reply store/store.go]: ok, fair

File: store/store.go (hunk 6 above)
Comment (chipaca) [reply store/store.go]: agreed! Any objections to shuffling the other uses of retry to be similar, in the same pr?

File: cmd/snap/complete.go (hunk 9 above)
Comment (chipaca) [reply cmd/snap/complete.go]: it's good enough well beyond that:
```
~$ wc -l /usr/share/dict/british-english-insane 
650656 /usr/share/dict/british-english-insane
~$ time /tmp/completeFromSortedFile /usr/share/dict/british-english-insane abo | wc -l
<nil>
190

real	0m0.025s
user	0m0.016s
sys	0m0.008s
```

the limit is somewhere in the 1M range:
```
~$ wc -l /var/tmp/out
10000001 /var/tmp/out
~$ time /tmp/completeFromSortedFile /var/tmp/out abo | wc -l
<nil>
2389706

real	0m5.282s
user	0m3.944s
sys	0m4.836s
```

but then we can bisect:

```
/tmp/pts-line-bisect$ time ./pts_lbsearch -pc /usr/share/dict/british-english-insane abo | wc -l
190

real	0m0.003s
user	0m0.000s
sys	0m0.000s
/tmp/pts-line-bisect$ time ./pts_lbsearch -pc /var/tmp/out abo | wc -l
2389706

real	0m0.041s
user	0m0.032s
sys	0m0.036s
```

given that the android and ios appstores both have less than 3M apps, I think having something that works well into that range of values is good enough, period.

File: cmd/snap/complete.go (hunk 10 above)
Comment (chipaca) [reply cmd/snap/complete.go]: I'll believe you on this one :-)

File: daemon/api_test.go (hunk 11 above)
Comment (chipaca) [reply daemon/api_test.go]: because the overlord manager's init now writes that variable

File: tests/main/completion/task.yaml (hunk 13 above)
Comment (chipaca) [reply tests/main/completion/task.yaml]: starting snapd created it. Maybe the test should stop snapd, move the old names aside, then do this, and restore it on restore?

File: dirs/dirs.go (hunk 14 above)
Comment (chipaca) [reply dirs/dirs.go]: gah. you are correct. See, this is (one of the reasons) why I'd originally had it under `/var/lib/snapd` -- laziness.

File: cmd/snap/complete.go (hunk 9 above)
Comment (chipaca) [reply cmd/snap/complete.go]: average times on a bbb with the 500k wordlist are ~.5s, which is a little too much but still faster than the remote completion we have today. With the 10M wordlist, it's ~2 minutes. Both are ~.05s with bisection, but can grow to be ~1m if there are too many matches. So when we reach that many entries we probably want to not print all matches!

File: overlord/snapstate/snapmgr.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (niemeyer) [reply overlord/snapstate/snapmgr.go]: It'll be worse if we get a lock out because something in the middle panics. This would prevent even the systemctl restart that would usually happen, and potentially render a device dead.

[the remark this answers is not in the document]
Comment (niemeyer) [reply overlord/snapstate/snapmgr.go]: Ah, there's another alternative as well: splitting the logic out into a function.

File: store/store.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (niemeyer) [reply store/store.go]: Ironically, the only use we're making today is about names! :P

File: store/store.go (hunk 6 above)
Comment (niemeyer) [reply store/store.go]: No objections since the PR is about to go in. It's usually bad to mix these changes in when it's still pending reviews, since the reviewer would then spend 5 minutes just looking at those lines to figure that there were no changes. :)

File: cmd/snap/complete.go (hunk 9 above)
Comment (niemeyer) [reply cmd/snap/complete.go]: Wow, nice numbers. Thanks a lot for the tests!

Do you have an idea of how many commands we're supposed to expect for the command-not-found use?

File: daemon/api_test.go (hunk 11 above)
Comment (niemeyer) [reply daemon/api_test.go]: Can we please have a comment? Seems tempting to move it elsewhere.

File: tests/main/completion/task.yaml (hunk 13 above)
Comment (niemeyer) [reply tests/main/completion/task.yaml]: Or just remove it again, which sounds easier.. the test just removed it anyway, and this is the one test that cares the most about this file being right.

File: dirs/dirs.go (hunk 14 above)
Comment (niemeyer) [reply dirs/dirs.go]: Yeah, to be honest I wouldn't mind having these files inside /var/lib/snapd either.

File: store/store.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (chipaca) [reply store/store.go]: :-) yes, for now yes

File: cmd/snap/complete.go (hunk 9 above)
Comment (chipaca) [reply cmd/snap/complete.go]: not more than an order of magnitude more
